### PR TITLE
Kh/var 1

### PR DESCRIPTION
**Pull request by Khatskevich**
This patchset contains VERY different commits.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -360,6 +360,40 @@ Notes:
...
+
+Notes:
+
+* Only zero-size arrays and maps are supported by now.
+* Default value may be inherited from an inner field with a default
+value or overridden. Example:
+```lua
+local schema = {
+    type = "record", name = "Frob", fields = {
+        { name = "foo", default = {f1=1, f2={f2_1=2}}, type =
```
Comment (Totktonada) [inline README.md]: I think we should follow our code style consistently: whitespaces around `=` in tables.

File: README.md [hunk 2]
```diff
@@ -360,6 +360,40 @@ Notes:
...
+Default values are substituted in two cases:
+1. during flattening if the fields are not presented in the data
+2. during unflattening and schema evolution in case the target schema
+   has extra fields with the default values
+
+Notes:
+
+* Only zero-size arrays and maps are supported by now.
+* Default value may be inherited from an inner field with a default
+value or overridden. Example:
```
Comment (Totktonada) [inline README.md]: I think we should mention the difference btw explicit null (`box.NULL`) and lack of a value (`nil`).

File: avro_schema/frontend.lua [hunk 3]
```diff
@@ -1,5 +1,13 @@
 -- The frontend.
--- Loads schema and generates IR.
+--
+-- Module overview:
+-- * `copy_schema` - validate and create schema
+-- * `copy_data` - validate a data by a schema; copy it
```
Comment (Totktonada) [inline avro_schema/frontend.lua]: Proposed wording: return copy of the data augmented with default values.

File: avro_schema/frontend.lua [hunk 4]
```diff
@@ -978,9 +942,48 @@ function debug_print_indent(msg)
     print(msg)
 end
 
+--
+-- Compiler generates code from the IR, which is more generic than
+-- a bare schema. IR defines a 'from' schema, a 'to' schema and a mapping.
```
Comment (Totktonada) [inline avro_schema/frontend.lua]: Generated code means lua code? It worth to state it here.

Don’t get the idea about being 'more generic' then a 'bare schema', can you expand it?

File: avro_schema/frontend.lua [hunk 5]
```diff
@@ -978,9 +942,48 @@ function debug_print_indent(msg)
...
+--
+--   { type = 'RECORD', nullable = nil/true, nested = {
+--           type = '__RECORD__',
+--           from = <schema>, to = <schema>, i2o = ?, o2i = ?, ... }}
+--
+--   { type = 'UNION', nested = {
+--           type = '__UNION__',
+--           from = <schema>, to = <schema>, i2o = ?, ... }}
+--
+-- Legend: n      — size of fixed
```
Comment (Totktonada) [inline avro_schema/frontend.lua]: You have used em dash here, but ascii dash below. It worth to restrict ourself to ascii within a source code.

File: README.md [hunk 6]
```diff
@@ -360,6 +360,40 @@ Notes:
...
+## Default values
+
+Default values are substituted in two cases:
+1. during flattening if the fields are not presented in the data
+2. during unflattening and schema evolution in case the target schema
+   has extra fields with the default values
+
+Notes:
+
+* Only zero-size arrays and maps are supported by now.
```
[the remark this answers is not in the document]
Comment (Khatskevich) [reply README.md]: Rewritten.

File: README.md (hunk 1 above)
Comment (Khatskevich) [reply README.md]: Rewritten.

File: avro_schema/compiler.lua [hunk 7]
```diff
@@ -1,3 +1,24 @@
+--
+-- This module generates `il` code for a given schema.
```
[the remark this answers is not in the document]
Comment (Khatskevich) [reply avro_schema/compiler.lua]: Rewritten.

File: avro_schema/compiler.lua [hunk 8]
```diff
@@ -1,3 +1,24 @@
...
+--
+-- Terminology:
+-- * `ipv` register which stores current position in an output
+--   buffer.
+-- * `ipo` offset from the current position in ibuf to the next
+--   piece of data to be transormed.
```
[the remark this answers is not in the document]
Comment (Khatskevich) [reply avro_schema/compiler.lua]: fixed

File: avro_schema/compiler.lua [hunk 9]
```diff
@@ -1,3 +1,24 @@
...
+-- * `code` Lua array, which contains a sequence of opcodes.
+-- * mode: c - check, x - transform, n - get next
+--  * CHECK - ensure an object at [$ipv + ipo] matches
```
[the remark this answers is not in the document]
Comment (Khatskevich) [reply avro_schema/compiler.lua]: No, that is correct.

File: avro_schema/compiler.lua [hunk 10]
```diff
@@ -1,3 +1,24 @@
...
+--    the IR (a shallow check, ex: in an array ignore contents)
+--  * TRANSFORM - convert object at [$ipv + ipo] according to the IR,
```
[the remark this answers is not in the document]
Comment (Khatskevich) [reply avro_schema/compiler.lua]: No, that is correct.

File: avro_schema/compiler.lua [hunk 11]
```diff
@@ -1,3 +1,24 @@
...
+--    ensure obuf has enough capacity, render results at [$0], and update $0.
+--  * GET NEXT - compute position of an object following [$ipv + ipo]
```
[the remark this answers is not in the document]
Comment (Khatskevich) [reply avro_schema/compiler.lua]: No, that is correct.

File: avro_schema/compiler.lua (hunk 9 above)
[the remark this answers is not in the document]
Comment (Khatskevich) [reply avro_schema/compiler.lua]: Rewritten.

File: avro_schema/compiler.lua [hunk 12]
```diff
@@ -1,3 +1,24 @@
...
+--
+-- Terminology:
+-- * `ipv` register which stores current position in an output
+--   buffer.
+-- * `ipo` offset from the current position in ibuf to the next
+--   piece of data to be transormed.
+-- * `code` Lua array, which contains a sequence of opcodes.
+-- * mode: c - check, x - transform, n - get next
+--  * CHECK - ensure an object at [$ipv + ipo] matches
+--    the IR (a shallow check, ex: in an array ignore contents)
```
[the remark this answers is not in the document]
Comment (Khatskevich) [reply avro_schema/compiler.lua]: Rewritten.

File: avro_schema/compiler.lua [hunk 13]
```diff
@@ -1,3 +1,24 @@
...
+--  * TRANSFORM - convert object at [$ipv + ipo] according to the IR,
+--    ensure obuf has enough capacity, render results at [$0], and update $0.
+--  * GET NEXT - compute position of an object following [$ipv + ipo]
+--    and store it back to $ipv.
```
[the remark this answers is not in the document]
Comment (Khatskevich) [reply avro_schema/compiler.lua]: Rewritten.

File: avro_schema/compiler.lua [hunk 14]
```diff
@@ -43,6 +64,12 @@ end
 local function is_record_or_union(s)
     return type(s) == 'table' and (s.type or 'record') == 'record'
 end
+
+-- Those tables contain a size of a field by its type.
+-- Negative size means that the field is complex and may contain
+-- a data inside of it. However this field requires `|size|`
+-- places in a msgpack array.
```
[the remark this answers is not in the document]
Comment (Khatskevich) [reply avro_schema/compiler.lua]: Rewritten a little. I suppose that the new version is clear enough.

File: avro_schema/compiler.lua [hunk 15]
```diff
@@ -1,3 +1,24 @@
+--
+-- This module generates `il` code for a given schema.
+--
+-- Terminology:
+-- * `ipv` register which stores current position in an output
```
[the remark this answers is not in the document]
Comment (Khatskevich) [reply avro_schema/compiler.lua]: fixed

File: avro_schema/compiler.lua [hunk 16]
```diff
@@ -71,9 +98,13 @@ local default_nullable_type_size = {
 
 local schema_width
 local schema_width_cache = setmetatable({}, weak_keys)
+--
 -- How many places in an array should be reserved for the structure.
--- In case of negative result, the number of required places in obuf can be
--- different from the returned number.
+-- In case of positive result, the same obuf size is required.
+-- In case of negative result:
+-- * abs(size) places in an msgpack array are required
```
[the remark this answers is not in the document]
Comment (Khatskevich) [reply avro_schema/compiler.lua]: No. That is correct.

File: avro_schema/frontend.lua (hunk 3 above)
Comment (Khatskevich) [reply avro_schema/frontend.lua]: ok. corrected.

File: avro_schema/frontend.lua (hunk 4 above)
Comment (Khatskevich) [reply avro_schema/frontend.lua]: Rewritten.

File: avro_schema/frontend.lua (hunk 5 above)
Comment (Khatskevich) [reply avro_schema/frontend.lua]: Fixed.

**Pull request merged**